src/runtimes/clients/AutoConfigurableDockerComposeClient.ts

```diff
@@ -0,0 +1,83 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as vscode from 'vscode';
+import { ext } from '../../extensionVariables';
+import { execAsync } from '../../utils/execAsync';
+import { AsyncLazy } from '../../utils/lazy';
+import { DockerComposeClient, IContainerOrchestratorClient } from '../docker';
+import { isDockerComposeClient } from '../OrchestratorRuntimeManager';
+import { AutoConfigurableClient } from './AutoConfigurableClient';
+
+interface ComposeConfig {
+    commandName: string;
+    composeV2: boolean;
+}
+
+export class AutoConfigurableDockerComposeClient extends DockerComposeClient implements AutoConfigurableClient {
+    private readonly composeConfigLazy = new AsyncLazy<ComposeConfig>(() => this.detectComposeConfig());
+
+    public constructor() {
+        super();
+        this.reconfigure();
+    }
+
+    public reconfigure(): void {
+        this.composeConfigLazy.clear();
+    }
+
+    public async slowConfigure(): Promise<void> {
+        const config = await this.composeConfigLazy.getValue();
+        this.commandName = config.commandName;
+        this.composeV2 = config.composeV2;
+    }
+
+    private async detectComposeConfig(): Promise<ComposeConfig> {
+        const config = vscode.workspace.getConfiguration('docker');
+
+        let composeCommand = config.get<string | undefined>('composeCommand');
+
+        if (composeCommand) {
+            // User has explicitly set a compose command, so we will respect it
+
+            let isComposeV2 = false;
+            if (/^docker(\s+compose\s*)?$/i.test(composeCommand)) {
+                // Normalize both "docker" and "docker compose" to "docker", with `isComposeV2` true
+                composeCommand = 'docker';
+                isComposeV2 = true;
+            }
+
+            return {
+                commandName: composeCommand,
+                composeV2: isComposeV2,
+            };
+        } else {
+            // User has not set a compose command, so we will attempt to autodetect it
+
+            try {
+                ext.outputChannel.appendLine('Attempting to autodetect Docker Compose command...');
+                await execAsync('docker compose version');

```

Comment:
Interested in others' thoughts on an idea I have. If we successfully detect Compose V2, we could stamp a memento, to avoid ever running this again, saving on perf. This makes the assumption that once a user installs a CLI version with Compose V2, they will never go back to one without it--if they do, they must use the setting to override the conclusion from that memento.

Reply:
How bad is the perf hit?

Reply:
Seemed like something in the vicinity of half a second, once per session, and only if you use compose features. Not bad.

Reply:
In that case my vote would be to err on the side of more robust/frequent detection